[MB 13164] PPM Closeout Expense Dropdown

pearl-truss opened the pull request:
## [Jira ticket](https://dp3.atlassian.net/browse/MB-13164) for this change

## Summary
This PR created the ExpenseForm component and displays different scenarios of its use in storybook under Customer Components / PPM Closeout / Expense Form

Storybook https://output.circle-artifacts.com/output/job/4d27fb3e-efb8-4c2f-8b3e-1608258c59bb/artifacts/0/storybook/index.html?path=/story/customer-components-ppm-closeout-expenses-ppm-form--sit-expenses

Designs https://app.zeplin.io/project/625896e702018f8ff82451c0/screen/625897c910f1e1907c1b2151

## Setup to Run Your Code

Start the Storybook locally.

```sh
make storybook
```

## Screenshots
<img width="973" alt="Screen Shot 2022-08-02 at 3 48 07 PM" src="https://user-images.githubusercontent.com/67110378/182460766-b0270e0b-cf03-44b9-bb54-129729c64143.png">
<img width="852" alt="Screen Shot 2022-08-02 at 3 48 18 PM" src="https://user-images.githubusercontent.com/67110378/182460771-0839d7d3-2121-474f-a56d-1305b3470e3d.png">
<img width="931" alt="Screen Shot 2022-08-02 at 3 48 27 PM" src="https://user-images.githubusercontent.com/67110378/182460773-36f837e2-8cc4-4b47-9770-a3e47c501574.png">
<img width="796" alt="Screen Shot 2022-08-03 at 11 00 34 AM" src="https://user-images.githubusercontent.com/67110378/182654854-15b45bfc-3d68-4d30-bdd3-5f19a8fdfa2f.png">

robot-mymove commented:
<!--
  0 failure: 
  2 warning:  This PR does not ..., Please add the JI...
  
  
  DangerID: danger-id-Danger;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This PR does not include changes to unit tests, even though it affects app code.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>Please add the JIRA issue key to the PR title (e.g. MB-123)</td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 3a7cba515019af32845a5db8d5710edae62d637a
</p>

dinabeigelman reviewed (commented):
Took a look and found a few things:
Should there be a char limit for the text input "What did you buy?"? I can just keep pasting more words in and I guess it's fine?

When you change the dates on the SIT options the calculation gets really wonky
<img width="533" alt="Screen Shot 2022-08-02 at 2 53 18 PM" src="https://user-images.githubusercontent.com/7771969/182480701-d62d949e-ebd3-4dd2-b3c1-2a47dc45d4dd.png">
<img width="510" alt="Screen Shot 2022-08-02 at 2 53 26 PM" src="https://user-images.githubusercontent.com/7771969/182480719-ab9dd40d-97a5-4d7c-84cf-1f7917573e25.png">

If you delete the already uploaded file and upload a new PDF, it looks like it was successful but gives a very strange error.
<img width="835" alt="Screen Shot 2022-08-02 at 2 54 06 PM" src="https://user-images.githubusercontent.com/7771969/182480942-731599f5-772b-4689-b974-d86aa2649e00.png">

Additionally, if you add a new PDF the styling is very different between what was uploaded previously and what was just uploaded. I'm not the most familiar with our file upload component so not sure if this is as designed.
<img width="762" alt="Screen Shot 2022-08-02 at 3 01 26 PM" src="https://user-images.githubusercontent.com/7771969/182481081-71b2b70e-2134-4e72-8dc7-8b85047f4ed3.png">

Lastly, I don't know if this is within the scope of this ticket, but I think @scarequotes and @AnnaGingle created an alert for when the user checks "I don't have my receipt" here: https://www.figma.com/file/tRkjtoWjT8727XH5m9k7mK/Milmove-Customer-App?node-id=305%3A1411

pearl-truss commented:
Thanks Dina! 

I've addressed your feedback, changes can be viewed with the [updated url](https://output.circle-artifacts.com/output/job/4d27fb3e-efb8-4c2f-8b3e-1608258c59bb/artifacts/0/storybook/index.html?path=/story/customer-components-ppm-closeout-expenses-ppm-form--sit-expenses)
- Quite a few of what you listed here are just storybook specific interaction issues. I went ahead and mocked out the interactions for file uploading and deleting since storybook can't use our actual APIs for handling these events. So if you were to click on the file upload component, you'll just see the event registered in the actions tab
- The extra decimal issue has been addressed 
-  FWIW we have other text fields in the PPM closeout flow (like the vehicle description field on the Weight Ticket screen) that aren't implementing a character limit
- Thanks for the link to the alert for what happens when the missing receipt button is checked. That was missing from this ticket. I went ahead and added it here. However, I noticed the blue used on figma is different than what's used for our Alert component. I'm not sure if that's an intentional change that should be introduced to all info type Alerts?

<img width="820" alt="Screen Shot 2022-08-03 at 11 55 55 AM" src="https://user-images.githubusercontent.com/67110378/182654510-b349bc0f-20e0-4bc1-bf60-c05a2f25cd27.png">

duncan-truss commented on src/utils/dates.js:
```diff
@@ -0,0 +1,10 @@
+export default function numOfDaysBetweenDates(date1, date2) {
+  const jsDate1 = new Date(date1);
+  const jsDate2 = new Date(date2);
```
It looks like the javascript Date constructor is setting a value for the hours and the difference in the daylight savings is causing the non-round value.  Moment might be easier to use here or [setting the hours value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/setHours) to zero might help reduce the variability.

![Screen Shot 2022-08-03 at 12 48 32 PM](https://user-images.githubusercontent.com/52669884/182664834-34d44669-9449-4adf-a94a-69856d89d2d9.png)

pearl-truss replied:
I thought there was an ADR to not use moment so I did this instead. But if it's okay to use moment I'll give that I try next (that would of been my first choice otherwise)

duncan-truss replied:
I think if we had an alternative already introduced like DayJS that would be the preference but the "standard" Javascript date library (or non-standard browser implementations) has it's own issues and the API is lacking.

ronaktruss replied:
Just sharing this because this tripped me up a bit:

Looking at the example screenshot that Duncan shared I was a little surprised that this wasn't working. The reason I would expect that to work is that the underlying time (that is retrieved when using getTime()) represents the beginning of the day in UTC time so the local timezones shouldn't have any impact.

After a closer look, it looks like the UI actually passes the dates using a slightly different format. For example the dates are passed in as "02 Sep 2022" or "07 Nov 2022" which when calling new Date on returns the beginning of the day in your local time.

duncan-truss replied:
Ahh good catch about the api vs field formats, I did print the app values out locally but then did something different when trying to illustrate the console example.

![Screen Shot 2022-08-03 at 3 22 05 PM](https://user-images.githubusercontent.com/52669884/182692134-416bbcd6-c1ee-4d26-bb20-973f93eaee84.png)

duncan-truss commented on src/utils/dates.js:
```diff
@@ -1,10 +1,8 @@
-  const timeDiff = jsDate2.getTime() - jsDate1.getTime();
+export default function numOfDaysBetweenDates(date1, date2) {
+  const mDate1 = moment(date1);
+  const mDate2 = moment(date2);
```
I think we may want `moment.utc(date)` so no local time conversion is applied https://momentjs.com/docs/#/parsing/utc/

duncan-truss commented on src/components/Customer/PPM/Closeout/ExpenseForm/ExpenseForm.jsx:
```diff
@@ -0,0 +1,238 @@
+                    <DatePickerInput name="sitEndDate" label="End date" />
+                    <h3>
+                      Days in storage:{' '}
+                      {values.sitStartDate && values.sitEndDate
```
You may want to also check that there are no errors for those SIT date fields. You can still end up with a NaN for the number of days if one of those fields has a type parse error.

duncan-truss commented on src/components/Customer/PPM/Closeout/ExpenseForm/ExpenseForm.test.jsx:
```diff
@@ -0,0 +1,223 @@
+      expect(screen.getByLabelText('Start date')).toHaveDisplayValue('24 Sep 2022');
+      expect(screen.getByLabelText('End date')).toBeInstanceOf(HTMLInputElement);
+      expect(screen.getByLabelText('End date')).toHaveDisplayValue('26 Dec 2022');
+
```
Could you add a check for the days in storage value being displayed?

duncan-truss commented on src/components/Customer/PPM/Closeout/ExpenseForm/ExpenseForm.jsx:
```diff
@@ -0,0 +1,238 @@
+    .when('expenseType', {
+      is: 'storage',
+      then: (schema) => schema.required('Required'),
+    }),
```
I think you could chain another when clause on here for the sitEndDate needing to be equal to or after the sitStartDate https://github.com/jquense/yup#schemawhenkeys-string--string-builder-object--values-any-schema--schema-schema

dinabeigelman commented:
@pearl-truss I took another look and I think it looks good!

I will admit I don't know what changes were made to the upload functionality but understand that there are limitations because of Storybook so will assume that when this is on a page in Milmove we can take another look there.

I'm checking with @AnnaGingle about a copy thing in the alert [here](https://www.figma.com/file/tRkjtoWjT8727XH5m9k7mK?node-id=305:1411#240505116) so once I hear back we can call this done. With regard to the Alert styling, yeah, it's very different in Storybook than in our designs. What we have in our designs should be the source of truth but because it takes engineer time to update the component in Storybook it's not a blocker at the moment.

pearl-truss commented:
@dinabeigelman does copy match? I pushed a recent change
<img width="771" alt="Screen Shot 2022-08-05 at 11 22 27 AM" src="https://user-images.githubusercontent.com/67110378/183109410-781ce8d0-3fc9-4d5d-88d8-1a17d4af14cc.png">

dinabeigelman commented:
@pearl-truss looks good to me!
